BackendTaskPilot/Program.cs

```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
+}
+
+public struct Rate
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
```

Comment:
Right now, the API call and data processing are handled together in a single function. To improve readability and separation of concerns, I’d suggest splitting the API call into its own dedicated method. You're also missing an `await` here. Whenever we’re working with async tasks, we should ensure proper error handling with a `try/catch` block to log issues or handle them gracefully.

For example:

```csharp
// We are extending from the previous comment about HttpFactory and adding a logger to the implementation
private readonly ILogger<ExchangeRateService> _logger;

public ExchangeRateService(HttpClient httpClient, ILogger<ExchangeRateService> logger)
{
    _httpClient = httpClient;
    _logger = logger;
}

private async Task<string> FetchRatesAsync()
{
    try
    {
        var response = await _httpClient.GetAsync(_baseUrl);

        if (!response.IsSuccessStatusCode)
        {
            _logger.LogError("Failed to fetch rates. Status code: {StatusCode}", response.StatusCode);
            throw new HttpRequestException($"Failed to fetch rates: {response.StatusCode}");
        }

        // Using `await` ensures proper handling of asynchronous tasks and avoids returning an unwrapped Task object.
        return await response.Content.ReadAsStringAsync();
    }
    catch (Exception ex)
    {
        _logger.LogError(ex, "An error occurred while fetching rates.");
        throw;
    }
}
```

This method would:
- Handle making the HTTP request.
- Validate the HTTP response (e.g., check for success status codes).
- Log any exceptions that occur during the API call with ILogger.
- Return the raw JSON string for further processing.

Separating this logic ensures the main function remains focused on higher-level tasks like processing or validating the data. It also makes the HTTP call easier to test in isolation. We should always keep testability in mind when creating a new service 🙂

Reply:
Additionally, we should really explore adding a cache here as we are going towards an outside APi where we may have to pay per request. It might be a bad idea if we need by the millisecond updated results, but usually they would publish some sort of guidelines on how often rates are changing. 

We also might run into rate limits if we throw a lot of data at them. I will say, however, that adding a cache prematurely or with incorrect TTL can be very damaging, as well. Care should be taken here, you should probably do this in a pair programming session 🙂 

Here’s what we could do:
1. Use **`IMemoryCache`** for in-memory caching to store exchange rates for a short period (e.g., 5–10 minutes, do some research around this particular API on what would be appropriate).
2. Cache the deserialized and filtered rates to avoid making repeated calls to the external API for the same data.
3. Ensure the cache is updated periodically to keep the data fresh.

For example, the logic could:
- Check the cache first. If data exists, return it.
- If not, fetch the data from the API, process it, and store it in the cache before returning it.

This approach would:
- Improve response times for repeat requests.
- Reduce the load on the external API.
- Make the service more scalable.

If you’d like help implementing this, I am always available on Slack for further questions or pair programming!